test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -239,7 +239,32 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
 		})
 		Expect(err).ToNot(HaveOccurred(), "Lifecycle hook calls were not as expected")
 
+		By("Deleting the workload cluster")
+		cluster := &clusterv1.Cluster{}
+		Eventually(func() error {
+			return input.BootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKey{Namespace: namespace.Name, Name: clusterName}, cluster)
+		}).Should(Succeed())
+
+		// Dump all the logs from the workload cluster before deleting them.
+		input.BootstrapClusterProxy.CollectWorkloadClusterLogs(ctx, cluster.Namespace, cluster.Name, filepath.Join(input.ArtifactFolder, "clusters", cluster.Name))
+
+		Byf("Dumping all the Cluster API resources in the %q namespace", namespace.Name)
+
+		// Dump all Cluster API related resources to artifacts before deleting them.
+		framework.DumpAllResources(ctx, framework.DumpAllResourcesInput{
+			Lister:    input.BootstrapClusterProxy.GetClient(),
+			Namespace: namespace.Name,
+			LogPath:   filepath.Join(input.ArtifactFolder, "clusters", input.BootstrapClusterProxy.GetName(), "resources"),
+		})
+
+		beforeClusterDeleteHandler(ctx, input.BootstrapClusterProxy.GetClient(), namespace.Name, clusterName, input.E2EConfig.GetIntervals(specName, "wait-machine-upgrade"))
```

Comment:
Looks like this is being called before a Delete is even issued on the Cluster therefore this will fail as the `BeforeClusterDelete` hook is never called.

Also you might not be abel to reuse `runtimeHookTestHandler` inside `BeforeClusterDelete` becasue `runtimeHookTestHandler` checks for `TopologyReconciled` condition to be false and blocking the Delete using BeforeClusterDelete hook does not update the `TopologyReconciled` condition - this is still pending discussion on who we want ti surface that information.